### PR TITLE
introduce KotlinResultCallAdapter for nice suspending network calls

### PR DESCRIPTION
**Pull request by connyduck**

### CODE REVIEW

File: app/src/main/java/com/keylesspalace/tusky/components/announcements/AnnouncementsViewModel.kt [hunk 1]
```diff
@@ -56,8 +57,9 @@ class AnnouncementsViewModel @Inject constructor(
             appDatabase.instanceDao().loadMetadataForInstance(accountManager.activeAccount?.domain!!)
                 .map<Either<InstanceEntity, Instance>> { Either.Left(it) }
                 .onErrorResumeNext {
-                    mastodonApi.getInstance()
-                        .map { Either.Right(it) }
+                    rxSingle {
```
Comment (charlag) [inline app/src/main/java/com/keylesspalace/tusky/components/announcements/AnnouncementsViewModel.kt:60]: honestly should just replace `mastodonApi.getCustomEmojis(),` with coroutines too and just remove this whole mess

File: app/src/main/java/com/keylesspalace/tusky/components/login/LoginActivity.kt [hunk 2]
```diff
@@ -166,32 +165,33 @@ class LoginActivity : BaseActivity(), Injectable {
...
-                .putString(DOMAIN, domain)
-                .putString(CLIENT_ID, credentials.clientId)
-                .putString(CLIENT_SECRET, credentials.clientSecret)
-                .apply()
-
-            redirectUserToAuthorizeAndLogin(domain, credentials.clientId)
+            mastodonApi.authenticateApp(
+                domain, getString(R.string.app_name), oauthRedirectUri,
+                OAUTH_SCOPES, getString(R.string.tusky_website)
+            ).fold(
```
Comment (charlag) [inline app/src/main/java/com/keylesspalace/tusky/components/login/LoginActivity.kt:171]: I must say I don't like using `fold` except for one very specific case: when you assign to a variable and provide default value. I would argue that `try/catch` is there for a reason.

File: app/src/main/java/com/keylesspalace/tusky/components/announcements/AnnouncementsViewModel.kt (hunk 1 above)
Comment (connyduck) [reply app/src/main/java/com/keylesspalace/tusky/components/announcements/AnnouncementsViewModel.kt:60]: I was planning to do this in another PR, but I can also do it with this one

File: app/src/main/java/com/keylesspalace/tusky/components/login/LoginActivity.kt (hunk 2 above)
Comment (connyduck) [reply app/src/main/java/com/keylesspalace/tusky/components/login/LoginActivity.kt:171]: But `fold` returns a value, you can do something like

```
val data  = mastodonApi.getData().fold(
    { // map data
    },
    { // return default value
    }
)
```

If you just want a default value you can do 

```
val data = mastodonApi.getData().getOrDefault(/* whatever default you need */)
```

File: app/src/main/java/com/keylesspalace/tusky/components/announcements/AnnouncementsViewModel.kt (hunk 1 above)
Comment (charlag) [reply app/src/main/java/com/keylesspalace/tusky/components/announcements/AnnouncementsViewModel.kt:60]: I think either way is fine

File: app/src/main/java/com/keylesspalace/tusky/components/login/LoginActivity.kt (hunk 2 above)
Comment (charlag) [reply app/src/main/java/com/keylesspalace/tusky/components/login/LoginActivity.kt:171]: Ooooh I got some language soup in my head, I'm sorry :sweat_smile: 
Of course it's `fold()`, it's a `Result`! 
Don't listen to me, everything is great.

**Pull request merged**